libraries/chain/db_market.cpp

```diff
@@ -367,18 +376,137 @@ bool database::apply_order(const limit_order_object& new_order_object, bool allo
+
+   bool to_check_call_orders = false;
+   const asset_object& sell_asset = sell_asset_id( *this );
+   const asset_bitasset_data_object* sell_abd = nullptr;
+   price call_match_price;
+   if( sell_asset.is_market_issued() )
+   {
+      sell_abd = &sell_asset.bitasset_data( *this );
+      if( sell_abd->options.short_backing_asset == recv_asset_id
+          && !sell_abd->is_prediction_market
+          && !sell_abd->has_settlement()
+          && !sell_abd->current_feed.settlement_price.is_null() )
+      {
+         call_match_price = ~sell_abd->current_feed.max_short_squeeze_price();
+         if( ~new_order_object.sell_price <= call_match_price ) // new limit order price is good enough to match a call
+            to_check_call_orders = true;
+      }
+   }
+
+   bool finished = false; // whether the new order is gone
+   if( to_check_call_orders )
+   {
+      // check limit orders first, match the ones with better price in comparison to call orders
+      while( !finished && limit_itr != limit_end && limit_itr->sell_price > call_match_price )
+      {
+         auto old_limit_itr = limit_itr;
+         ++limit_itr;
+         // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+         finished = ( match( new_order_object, *old_limit_itr, old_limit_itr->sell_price ) != 2 );
+      }
+
+      if( !finished )
+      {
+         // check if there are margin calls
+         const auto& call_price_idx = get_index_type<call_order_index>().indices().get<by_price>();
+         auto call_min = price::min( recv_asset_id, sell_asset_id );
+         auto call_itr = call_price_idx.lower_bound( call_min );
+         // feed protected https://github.com/cryptonomex/graphene/issues/436
+         auto call_end = call_price_idx.lower_bound( ~sell_abd->current_feed.settlement_price );
+         while( !finished && call_itr != call_end )
+         {
+            auto old_call_itr = call_itr;
+            ++call_itr; // would be safe, since we'll end the loop if a call order is partially matched
+            // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+            // assume hard fork core-338 and core-625 will take place at same time, not checking HARDFORK_CORE_338_TIME here.
+            finished = ( match( new_order_object, *old_call_itr, call_match_price ) != 2 );
+         }
+      }
+   }
+
+   // still need to check limit orders
+   while( !finished && limit_itr != limit_end )
+   {
+      auto old_limit_itr = limit_itr;
+      ++limit_itr;
+      // match returns 2 when only the old order was fully filled. In this case, we keep matching; otherwise, we stop.
+      finished = ( match( new_order_object, *old_limit_itr, old_limit_itr->sell_price ) != 2 );
+   }
+
+   const limit_order_object* updated_order_object = find< limit_order_object >( order_id );
+   if( updated_order_object == nullptr )
```

Comment:
This code is copied from the old `apply_order` function. I was wondering the same thing.

Actually, due to https://github.com/cryptonomex/graphene/issues/555 ("Defer something-for-nothing culling for taker orders until the order is unmatched") and fix for #184 (taker order may suffer something-for-nothing issue), `finished == true` here only implies that the new order is unable to match, it can be filled already, or be too small so unable to fill.